src/onnx/parse_mean_variance_norm.cpp

```
@@ -0,0 +1,94 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#include <migraphx/onnx/op_parser.hpp>
+#include <migraphx/make_op.hpp>
+#include <migraphx/instruction.hpp>
+#include <migraphx/ranges.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+namespace onnx {
+
+struct mean_variance_norm : op_parser<mean_variance_norm>
+{
+    std::set<shape::type_t> valid_types = {
+        shape::bf16_type, shape::double_type, shape::float_type};
+
+    std::vector<op_desc> operators() const
+    {
+        return {{"MeanVarianceNormalization", "mean_variance_norm"}}; 
+    }
+
+    instruction_ref parse(const op_desc& opd,
+        const onnx_parser& parser,
+        const onnx_parser::node_info& info,
+        std::vector<instruction_ref> args) const
+    {
+        const auto dtype         = args[0]->get_shape().type();
+        const auto literal_dtype = dtype;
+
+        if(not contains(valid_types, dtype))
+        {
+            MIGRAPHX_THROW(opd.onnx_name + ": invalid output type: " + std::to_string(dtype) +
+                           ". Valid types are (bfloat16), (double), and (float).");
+        }
+
+        const auto& X                  = args[0];
+
+        const auto eps_default            = 1e-9f;
+        const auto axes_default = std::vector<size_t>{0, 2, 3};
+        
+        auto eps = eps_default;
+        if (contains(info.attributes, "epsilon"))
+        {
+            eps = parser.parse_value(info.attributes.at("epsilon")).at<float>();
+        }
+
+        auto axes = axes_default;
+        auto axes_min_size = axes.size();
+        if (contains(info.attributes, "axes"))
+        {
+            axes.assign(info.attributes.at("axes").ints().begin(), info.attributes.at("axes").ints().end());
+            axes_min_size = axes.size();
+        }
+        assert(X->get_shape().ndim() >= axes_min_size);
```

In the parser you should throw an exception when a condition isn't met. You can do this using the MIGRAPHX_THROW macro which takes a string as input. There isn't a set format for the string, I usually go with "OperatorName: input_dim/attribute/anything has value x, should be y", something like you have above for the type validty check.

---

Fair enough... note: see parse_instancenorm.cpp:88  there's also an assert(.) call.